R/string_boundary_linter.R

```diff
@@ -0,0 +1,86 @@
+#' Require usage of startsWith() and endsWith() over grepl()/substr() versions
+#'
+#' [startsWith()] is used to detect fixed initial substrings; it is more
+#'   readable and more efficient than equivalents using [grepl()] or [substr()].
+#'   c.f. `startsWith(x, "abc")`, `grepl("^abc", x)`,
+#'   `substr(x, 1L, 3L) == "abc"`.
+#'
+#' Ditto for using [endsWith()] to detect fixed terminal substrings.
+#'
+#' @evalRd rd_tags("string_boundary_linter")
+#' @seealso [linters] for a complete list of linters available in lintr.
+#' @export
+string_boundary_linter <- function() {
+  str_cond <- xp_and(
+    "string-length(text()) > 3",
+    "contains(text(), '^') or contains(text(), '$')"
+  )
+  grepl_xpath <- glue::glue("//expr[
+    expr[SYMBOL_FUNCTION_CALL[text() = 'grepl']]
+    and not(SYMBOL_SUB[
+      text() = 'ignore.case'
+      and not(following-sibling::expr[1][NUM_CONST[text() = 'FALSE']])
+    ])
+    and not(SYMBOL_SUB[
+      text() = 'fixed'
+      and not(following-sibling::expr[1][NUM_CONST[text() = 'FALSE']])
```

Comment:
ditto

Reply:
not sure what you mean here

Reply:
The `F` check.

Reply:
so `NUM_CONST[1]`? but there can only be one `NUM_CONST` without further nesting. So again, a bit redundant, and not consistent with XPaths elsewhere in lintr.

Reply:
I just meant that a second check for F was needed at this location.

Reply:
oh I see... already resolved then. Mixed up across review cycles :)